### PR TITLE
[flang][driver] Make -stdlib= option visible to flang and silently ignored by it

### PR DESCRIPTION
**Pull request by pawosm-arm**
This patch is to address a sudden problem I have encountered while trying to build mixed-language Fortran/C++ project. As I wanted the C++ part to utilize libc++ (instead of system-provided libstdc++), I've specified the `-stdlib=libc++` option globally, in the CXXFLAGS and LDFLAGS variables. This has saved me from making expensive changes in the Makefiles. Unfortunately, since it was the flang complier that was invoked for linking, suddenly it ended up with an error:

```
flang-new: error: unknown argument '-stdlib=libc++'; did you mean '-Xclang -stdlib=libc++'?
```

This has created a nowhere to go situation, the only remaining option is to make the -stdlib flag visible to flang and silently ignored.

### CODE REVIEW

**Comment by tblah** [comment]
I don't think it is correct to say that the option was ignored. It may still be read by the compiler driver (which generates the link line). In fact it probably *should* be processed by the compiler driver in order to set the C++ standard library correctly - otherwise you could just not set this in your `LDFLAGS`.

File: flang/test/Integration/mixed-lang-stdlib.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
+// REQUIRES: system-linux
```
Comment (tblah) [inline flang/test/Integration/mixed-lang-stdlib.cpp:1]: I'm not sure about this test. Usually we would go about testing something like this by running FileCheck on the output of `flang-new -###` to see what is passed to the linker (and that the linker gets called at all instead of us generating an error for an invalid argument).

Comment (pawosm-arm) [reply flang/test/Integration/mixed-lang-stdlib.cpp:1]: As I stated in my other comment, the `-###` output is not affected by the `-stdlib=` flag, so a test which looks at its output would not provide a great value. The integration test from the other hand guards that the real-life scenario that I've encountered could be resolved with `flang-new`.

**Comment by tblah** [comment]
If this is completely ignored and does not effect the link line, are you sure the resulting binary actually uses the right c++ library?

**Comment by tblah** [comment]
We discussed this further offline and I am now convinced there could be a valid usecase for this. Some C++ code might use header-only parts of the C++ standard library. Setting `-stdlib=` will effect the include paths clang uses for stdlib when compiling.

When flang links the program it will not link to a C++ library by default so it is correct that this does nothing. But if cmake wants to include CXXFLAGS inside of LDFLAGS, it is more user friendly to allow this flag.

If the user uses something that would require linking to the C++ library they will get a linker error not silent linking to the wrong c++ library.

**Review by tblah — APPROVED** [review]
LGTM but if possible please wait for another reviewer

**Comment by pawosm-arm** [comment]
> If this is completely ignored and does not effect the link line, are you sure the resulting binary actually uses the right c++ library?
> 
> Edit: flang will not link a C++ library. It is not a C++ compiler.

As we discussed it offline, in case of this project it does not matter, no calls to the library are actually being made from the C++ code, and it was more important what CMake did with CXXFLAGS (so the headers from the right C++ standard library were used). Also the integration test I've added depicts that flang-new doesn't really need to know anything about the language and standard libraries they use. The trouble here is that linker flags were blindly propagated by CMake (when asked to specify C++ standard library) no matter if it has any use or not. In smaller project, where we know that C++ the part isn't actually calling anything from library, the -stdlib flang should not be used during linking at all, but here we're talking about large scale project, where the Flang/C++ part is just a small part so the project maintainers didn't really care about making an exception in the flags being used when linking just that part.

**Comment by Meinersbur** [comment]
While the FortranRuntime is written in C++, it is ensured that it does not depend on libc++.so/libstdc++.so ([regression test](https://github.com/llvm/llvm-project/blob/main/flang/test/Runtime/no-cpp-dep.c)). Hence what C++ runtime the user links into their executable shouldn't concern the Fortran part.

When mixing Fortran and C++ code with Fortran code in the same add_library/add_executable, will use the C++ linker to link all the object files together (just as with C and C++). That might be the expected thing to do. That means it will care for linking the correct C++ library, but leave out the Fortran runtime library, and the user has to add `-lFortranRuntime` manually. If I understand correct, with this patch `flang-new -stdlib=libc++` will link the C++ and Fortran standard libraries? Unfortunately CMake does not know that.

Note that the test will fail when `LLVM_DEFAULT_TARGET_TRIPLE` is set to anything else than the host platform. Because of this. Clang tests never invoke the linker, but check the output of `-###`. However, Flang already has tests with the same issue.

**Comment by pawosm-arm** [comment]
> Note that the test will fail when `LLVM_DEFAULT_TARGET_TRIPLE` is set to anything else than the host platform. Because of this. Clang tests never invoke the linker, but check the output of `-###`. However, Flang already has tests with the same issue.

I wonder how the other C++ integration test in this directory (`iso-fortran-binding.cpp`) is not affected, it is not only linking but also executing what has been linked, which I wanted to avoid, I didn't want to run into the issues with RPaths.

I also wanted to bring your attention to the subject in context of your work on extracting Fortran Runtime, what I've noticed is that using `-DLLVM_ENABLE_LIBCXX=True` CMake flag has unexpected effect when it comes to C++/Fortran mixed code.

**Comment by pawosm-arm** [comment]
> (and there isn't even a warning that the option is unused)
> 
> @pawosm-arm Wasn't this ever an issue? Should those to be consistent?

Interesting, clang would do the same as flang-new patched with my patch. Maybe addition of the warning would be a good idea, if it isn't an overkill.

**Comment by banach-space** [comment]
We should avoid dummy flags like this - they are usually an indication of a larger issue. 

I don't want to block this and, as @Meinersbur points out, there are some "inherited issues" to consider too. However, I'd like to identify the right long-term solution here. If possible.

> I've specified the -stdlib=libc++ option globally, in the CXXFLAGS and LDFLAGS variables

Why are CXXFLAGS used when invoking a Fortran compiler? That's not correct, is it? 

`LDFLAGS` would make sense, but then `-stdlib=` should be treated as a linker flag and `flang-new` as a linker driver. In fact, it sounds like `-stdlib=` could fall into some special category:
* valid when **compiling** C++,
* invalid when **compiling** Fortran,
* valid when **linking** mixed C++ and Fortran object files.

To me it sounds like:
*  flags from `CXXFLAGS` should not be used in Fortran _compilation_ flows,
* `-stdlib=` should be flagged as a linker option (that seems to be missing?),
* `flang-new` should allow/ignore "linker" flags that are not "visible" in Fortran _compilation_ flows (as in, the "Visibility" logic could be tweaked).

If `-stdlib=` were to be made available in Fortran compilation, the help-text should be updated accordingly.

**Comment by Meinersbur** [comment]
From the summary?

> This has created a nowhere to go situation, the only remaining option is to make the -stdlib flag visible to flang and silently ignored.

Can you clarify? I think it is not ignored, otherwise the test would fail since it requires linking to the `<iostream>` implementation. It must add `-l(std)c++` to the linker command line.


> > (and there isn't even a warning that the option is unused)
> > @pawosm-arm Wasn't this ever an issue? Should those to be consistent?
> 
> Interesting, clang would do the same as flang-new patched with my patch. Maybe addition of the warning would be a good idea, if it isn't an overkill.

AFAICS your patch only makes the `-stdlib` recognized by Flang. `lc++` being added must have been inherited from the C++ driver, but seemingly missing from the C driver path. 
A lot of flags are purposefully by the driver because build systems typically pass the same flags to every build phase which would otherwise cause errors ([example](https://github.com/llvm/llvm-project/issues/88173)). A warning usually only annoys people that compile with `-werror`.

> what I've noticed is that using -DLLVM_ENABLE_LIBCXX=True CMake flag has unexpected effect when it comes to C++/Fortran mixed code.

What effects?

**Comment by DavidTruby** [comment]
I missed replying on this review somehow, sorry to come back 2 weeks later...

> Why are CXXFLAGS used when invoking a Fortran compiler? That's not correct, is it?

I don't think CXXFLAGS will be being added to the flang invocation in this case. What I believe is happening is that cmake prefers to use the compiler driver for the relevant language to link the final binaries, so in this case is invoking `flang` as the _linker_ and passing it the LDFLAGS. 
This is why if you want to pass _actual_ linker options to the actual linker, you need to prefix them with `-Wl` in LDFLAGS so that the compiler driver forwards them on. E.g. `-Wl,--gc-sections` or whatever.

The CXXFLAGS are only passed to CXX and the FCFLAGS only passed to FC, and the LDFLAGS will be passed to whichever cmake chooses to use as the linker at the end _as well as_ CXXFLAGS or FCFLAGS depending which was picked. So probably you shouldn't be adding `-stdlib=libc++` to LDFLAGS only to CXXFLAGS. 

A quick test shows me that CMake picks the CXX compiler in mixed-language CXX and Fortran programs though, so I'm not sure why @pawosm-arm is seeing an error. It looks like CMake even does the right thing and manually adds -lFortranRuntime etc to the `clang++` link line when linking multi-language projects with both C++ and Fortran where CMAKE_Fortran_COMPILER=flang-new.

tl;dr this looks like a cmake misconfiguration issue to me?

**Comment by DavidTruby** [comment]
I don’t personally think that the flang driver should ever attempt to link a C++ library of any kind. FWIW while there’s no stdlib option to consider in their case, g++ will never auto link the gfortran runtimes and gfortran will never auto link libstdc++

Mixed C++/Fortran programs are always going to be complicated for exactly this reason: as a user I wouldn’t expect a C++ compiler driver to automatically link a Fortran runtime or vice versa, and would find it quite surprising if that happened. In my opinion the current behaviour (which matches g++/gfortran) to require build systems to handle the complexity of adding the right linkage flags is the correct one. This would leave us rejecting stdlib= as an unrecognised option on flang as is the case in current main.

> I don't think we can rely on every build system doing that, so I would still advocate that -stdlib=libc++ is not just ignored, but actually adds the C++ standard library linker flags (see https://github.com/llvm/llvm-project/pull/110598#issuecomment-2388005207) to support such language-combination scenarios

If we were to do things this way, what should flang do if no -stdlib flag is provided? Should it link the default C++ stdlib as clang++ does, or break with clang++ and not link any C++ stdlib if nothing is provided? Both would be surprising behaviour in my opinion, and both would also risk breaking the existing cmake support.

**Comment by Meinersbur** [comment]
> I don’t personally think that the flang driver should ever attempt to link a C++ library of any kind. FWIW while there’s no stdlib option to consider in their case, g++ will never auto link the gfortran runtimes and gfortran will never auto link libstdc++

gfortran doesn't have a flag that means "use this C++ standard library". Without this PR, Flang doesn't either which would be fine with me. OP would need to fix their build system.

With this PR, `flang --help` will list an option `-stdlib` with the help text "C++ standard library to use". But it doesn't do anything. It feels like it would be ignoring `-lc++`.


> If we were to do things this way, what should flang do if no -stdlib flag is provided? Should it link the default C++ stdlib as clang++ does, or break with clang++ and not link any C++ stdlib if nothing is provided? Both would be surprising behavior in my opinion, and both would also risk breaking the existing cmake support.

No `-stdlib` means no linking to any C++ standard library for Flang, no change of behavior without the flag. It also does not break cmake support since it never passes `-stdlib` by itself. Just asking for the flag's behavior to match its description.

**Comment by DavidTruby** [comment]
Sorry I wasn’t very clear in my previous comments; I think we shouldn’t go ahead with this PR and should continue to reject the flag. So I think we’re in agreement.

**Review by Meinersbur — CHANGES_REQUESTED** [review]
Marking this PR as no consensus reached

**Comment by DavidTruby** [comment]
@pawosm-arm do you still need this or do the CMake configuration changes I suggested work? Can we close this if it's not needed anymore?

**Comment by pawosm-arm** [comment]
There was a change of plans, so I don't need it now. But I expect the subject will return sooner or later raised by someone else who stumble upon this just like me.

Closing it for now.
